File: code/http/http_server.py
import socket


def handle_request(request):
    lines = request.split('\r\n')
    if not lines:
        return 'HTTP/1.1 400 Bad Request\r\n\r\n'

    method, path, _ = lines[0].split(' ', 2)
    print(f'  {method} {path}')

    body = f'''<!DOCTYPE html>
<html>
<head><title>Hello Network</title></head>
<body>
<h1>Hello Network!</h1>
<p>方法: {method}</p>
<p>路径: {path}</p>
<hr>
<h2>请求头</h2>
<pre>'''
    for line in lines[1:]:
        if line == '':
            break
        body += line + '\n'

    body += '''</pre>
</body>
</html>'''

    response = (
        'HTTP/1.1 200 OK\r\n'
        'Content-Type: text/html; charset=utf-8\r\n'
        f'Content-Length: {len(body.encode())}\r\n'
        'Connection: close\r\n'
        '\r\n'
        f'{body}'
    )
    return response


def main():
    host = '127.0.0.1'
    port = 8888

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind((host, port))
    server.listen(5)

    print(f'HTTP Server 启动于 http://{host}:{port}')
    print('按 Ctrl+C 停止\n')

    try:
        while True:
            conn, addr = server.accept()
            print(f'新连接: {addr[0]}:{addr[1]}')
            data = conn.recv(4096)
            if data:
                request = data.decode()
                response = handle_request(request)
                conn.sendall(response.encode())
            conn.close()
            print(f'关闭连接: {addr[0]}:{addr[1]}\n')
    except KeyboardInterrupt:
        print('\n服务器停止')
    finally:
        server.close()


if __name__ == '__main__':
    main()

File: code/http/HttpServer.java
import java.io.*;
import java.net.*;

/**
 * 简单的 HTTP 服务器示例
 * 展示如何通过 TCP Socket 处理 HTTP 请求
 *
 * 运行方式:
 *   javac HttpServer.java && java HttpServer
 *   浏览器访问 http://127.0.0.1:9999
 */
public class HttpServer {

    public static void main(String[] args) throws IOException {
        int port = 9999;
        ServerSocket server = new ServerSocket(port);
        System.out.println("HTTP Server 启动于 http://127.0.0.1:" + port);
        System.out.println("按 Ctrl+C 停止\n");

        while (true) {
            try (Socket client = server.accept();
                 BufferedReader in = new BufferedReader(
                         new InputStreamReader(client.getInputStream()));
                 PrintWriter out = new PrintWriter(
                         client.getOutputStream(), true)) {

                String requestLine = in.readLine();
                if (requestLine == null) continue;
                System.out.println("请求: " + requestLine);

                String body = "<!DOCTYPE html><html><head><meta charset='utf-8'>" +
                        "<title>Hello Network</title></head><body>" +
                        "<h1>Hello Network!</h1>" +
                        "<p>请求: " + requestLine + "</p>" +
                        "<hr><h2>请求头</h2><pre>";

                String line;
                while ((line = in.readLine()) != null && !line.isEmpty()) {
                    body += line + "\n";
                }
                body += "</pre></body></html>";

                out.print("HTTP/1.1 200 OK\r\n" +
                         "Content-Type: text/html; charset=utf-8\r\n" +
                         "Content-Length: " + body.getBytes().length + "\r\n" +
                         "Connection: close\r\n" +
                         "\r\n" +
                         body);
                out.flush();
            } catch (Exception e) {
                System.out.println("处理请求出错: " + e.getMessage());
            }
        }
    }
}
